fl-server/src/serve_flists.rs

```diff
@@ -77,25 +85,53 @@ pub async fn serve_flists(req: Request<Body>) -> impl IntoResponse {
     };
 }
 
-pub async fn visit_dir_one_level(path: &std::path::Path) -> io::Result<Vec<FileInfo>> {
+pub async fn visit_dir_one_level(
+    path: &std::path::Path,
+    state: &Arc<config::AppState>,
+) -> io::Result<Vec<FileInfo>> {
     let mut dir = tokio::fs::read_dir(path).await?;
     let mut files: Vec<FileInfo> = Vec::new();
 
     while let Some(child) = dir.next_entry().await? {
-        let the_uri_path = child.path().to_string_lossy().to_string();
+        let path_uri = child.path().to_string_lossy().to_string();
+        let is_file = child.file_type().await?.is_file();
+        let name = child.file_name().to_string_lossy().to_string();
+
+        let mut progress = 0.0;
+        if is_file {
+            match state.flists_progress.lock().unwrap().get(&format!(
+                "{}/{}",
+                path.to_string_lossy().to_string(),
+                name
+            )) {
+                Some(p) => progress = p.to_owned(),
+                None => progress = 100.0,
+            }
+
+            let ext = child
+                .path()
+                .extension()
+                .unwrap()
```

Comment:
Is it safe to do unwrap here? are you sure there is noway this going to fail? if failure is a possible scenario here then maybe it's better to check and return error instead of crashing the process